File: ctci/removeDuplicates.js
(function(){
  'use strict';
  /*
  2.1 Write code to remove duplicates from an unsorted linked list.
FOLLOW UP
How would you solve this problem if a temporary buffer is not allowed?
   */
  function removeDuplicates(head){
    // traverse through linked list
    // add unique elements into an object
    var current = head;
    var previous;
    if (!current) return head;
    else {
      var uniques = {};
      while(current.next){
        previous = current;
        current = current.next;
        if (uniques[current.data] !== undefined){
          previous.next = current.next;
        }
        else {
          uniques[current.data] = 1;
        }
      }
    }
    return head;
  }

  // without temporary buffer

  function removeDuplicatesSort(head){
    // sort the linked list
    // traverse through the list, when u come across a chain (2 or more) duplicate values, remove those chains
    
  }
}());

File: python/removeDuplicates.py
class ListNode:
  def __init__(self, value):
    self.value = value
    self.next = None

def removeDuplicates(head):
  if head == None:
    return None
  if head.next == None:
    return head

  current = head.next
  prev = head
  list_values = dict()
  list_values[prev.value] = 1
  while current != None:
    if list_values.has_key(current.value):
      prev.next = current.next
    else:
      list_values[current.value] = 1
      prev = prev.next
    current = current.next
  return head

head = ListNode(0)
head.next = ListNode(1)
head.next.next = ListNode(1)
head.next.next.next = ListNode(2)
head.next.next.next.next = ListNode(1)

def printList(head):
  if head == None:
    return None
  print(head.value)
  return printList(head.next)

printList(head)
printList(removeDuplicates(head))